client/cmd/job/plan.go

```diff
@@ -0,0 +1,242 @@
+	}
+
+	planCmd.inject(cmd)
+	return cmd
+}
+
+func (p *planCommand) inject(cmd *cobra.Command) {
+	cmd.Flags().StringVarP(&p.configFilePath, "config", "c", config.EmptyPath, "File path for client configuration")
+	cmd.Flags().StringVarP(&p.output, "output", "o", "./plan.json", "File path for output of plan")
+	cmd.Flags().BoolVarP(&p.verbose, "verbose", "v", false, "Print details related to operation")
+
+	cmd.Flags().StringVar(&p.gitProvider, "git-provider", os.Getenv("GIT_PROVIDER"), "selected git provider used in the repository")
+	cmd.Flags().StringVar(&p.gitURL, "git-host", os.Getenv("GIT_HOST"), "Git host based on git provider used in the repository")
+	cmd.Flags().StringVar(&p.gitToken, "git-token", os.Getenv("GIT_TOKEN"), "Git token based on git provider used in the repository")
+	cmd.Flags().StringVar(&p.gitProjectID, "git-project-id", os.Getenv("GIT_PROJECT_ID"), "Determine which git project will be checked")
+	cmd.Flags().StringVar(&p.sourceRef, "source", p.sourceRef, "Git diff source reference (new state) [commit SHA, branch, tag]")
+	cmd.Flags().StringVar(&p.targetRef, "target", p.targetRef, "Git diff target reference (current state) [commit SHA, branch, tag]")
+}
+
+func (p *planCommand) PreRunE(_ *cobra.Command, _ []string) error {
+	var err error
+	p.logger.Info("[plan] compare: `%s` ← `%s`", p.targetRef, p.sourceRef)
+
+	p.clientConfig, err = config.LoadClientConfig(p.configFilePath)
+	if err != nil {
+		return err
+	}
+
+	switch p.gitProvider {
+	case "gitlab":
+		p.repository, err = gitlab.NewAPI(p.gitURL, p.gitToken)
+	case "github":
+		p.repository, err = github.NewAPI(p.gitURL, p.gitToken)
+	default:
+		return errors.New("unsupported git provider, we currently only support: [github,gitlab]")
+	}
+	if err != nil {
+		return err
+	}
+
+	p.specReadWriter, err = specio.NewJobSpecReadWriter(afero.NewOsFs())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *planCommand) RunE(_ *cobra.Command, _ []string) error {
+	var (
+		ctx   = context.Background()
+		plans plan.Plan
+		err   error
+	)
+
+	plans, err = p.generatePlanWithGitDiff(ctx)
+	if err != nil {
+		return err
+	}
+
+	p.printPlan(plans)
+	return p.savePlan(plans)
+}
+
+func (p *planCommand) generatePlanWithGitDiff(ctx context.Context) (plan.Plan, error) {
+	plans := plan.NewPlan(p.clientConfig.Project.Name)
+	affectedDirectories, err := p.getAffectedDirectory(ctx)
+	if err != nil {
+		return plans, err
+	}
+
+	for _, directory := range affectedDirectories {
+		namespace, err := p.getNamespaceNameByJobPath(directory)
+		if err != nil {
+			return plans, err
+		}
+
+		sourceSpec, err := p.getJobSpec(ctx, filepath.Join(directory, jobFileName), p.sourceRef)
+		if err != nil {
+			return plans, err
+		}
+		targetSpec, err := p.getJobSpec(ctx, filepath.Join(directory, jobFileName), p.targetRef)
+		if err != nil {
+			return plans, err
+		}
+
+		plans.Job.Add(namespace, sourceSpec.Name, targetSpec.Name, &plan.JobPlan{})
+	}
+
+	plans = plans.GetResult()
```

Comment:
Why are we assigning result of a plan to plan itself ?

Reply:
because in the GetResult it will merge create and delete operation as migrate if it has different namespace. How about I just returned it directly ?
```go
      return plans.GetResult(), nil
```